### ISSUE TITLE
Annotation `javax.annotation.processing.Generated` is used when building with jdk11 despite `targetVersion=8`

### ISSUE THREAD

**Issue by XN137** [joelittlejohn/jsonschema2pojo#1473]
we are building on jdk11 (but target a java 8 runtime) with the maven plugin and setting `<targetVersion>8</targetVersion>`

but getting the following compilation errors:
```
error: cannot find symbol
    symbol:   class Generated
    location: package javax.annotation.processing
```

so we have to add `jsonschema2pojo-jdk-annotation` to the classpath to workaround the problem.

however: when targetting java 8 it should not try to add the java9 annotation here imo:
https://github.com/joelittlejohn/jsonschema2pojo/blob/284c1e81556eaa0bd4515a186f25fc875384eef9/jsonschema2pojo-core/src/main/java/org/jsonschema2pojo/util/AnnotationHelper.java#L42-L46

i am guessing `Class.forName` works for the java9 classname because we are building on jdk11.

a more general solution that would also cover https://github.com/joelittlejohn/jsonschema2pojo/issues/1461 is to make the annotation class freely configurable with another mojo property like `generatedAnnotationClass`

it could default to `"auto"` and keep running the existing logic if desired

**Comment by joelittlejohn** [comment]
One issue with this having this as a configurable property is that most people don't care about this annotation, they just want something that works without receiving an error on compilation the first time they use this jsonschema2pojo. That said, there are a lot of benefits to having a Generated annotation - I removed it for a while because of these issues, but found that there's quite a lot of tooling that uses that annotation to decide how to treat our generated source code, so I added it back.

So assuming we keep it, it would be good to have both an unobtrusive and sensible default, as well as a config option for those that care to use it.

So I guess the logic should be:

1. `generationAnnotationClass` is specified: use that
2. `generationAnnotationClass` is not specified:
   * targetVersion = 8 or lower, use `javax.annotation.Generated`
   * targetVersion = 9 or higher, use `jakarta.annotation.Generated` is it is found, or if not then use `javax.annotation.processing.Generated` if it is found, or if not then use `javax.annotation.Generated` if it is found.

I think I'll add `jakarta.annotation.Generated` to jsonschema2pojo-jdk-annotation as well, as it can make life easier.

It's so frustrating that something so basic is not part of the Java's core.

**Comment by XN137** [comment]
@joelittlejohn thanks for the quick reply, i've crafted a small PR that would fix the intermediate problem:
https://github.com/joelittlejohn/jsonschema2pojo/pull/1474

please approve the workload + "review" if you have time

i lack a lot of context on why `jsonschema2pojo-jdk-annotation` was added in the first place but imo it is an anti-pattern to ship replacement-annotations of different libraries.
maybe there was a special reason for doing this in an android context but for plain java usage i think the bug i am describing here (not respecting `targetVersion` and the java9 class being available on jdk11) was what was causing the majority of problems originally...

**Comment by unkish** [comment]
Looks like a duplicate of https://github.com/joelittlejohn/jsonschema2pojo/issues/1361

**Comment by XN137** [comment]
@unkish you are right, i even commented on that ticket a while ago already :sweat_smile:
maybe with the suggest PR we can close both, lets see what @joelittlejohn says

**Comment by unkish** [comment]
To be honest, I'd rather see the issue solved properly.
Also: the date of next release is not sent in stone so hasting doesn't seem reasonable.

**Comment by XN137** [comment]
what does "properly" mean for you here?
using java 9 annotations when targetVersion=8 is clearly a bug imo...

whether you want to rework the whole annotation mechanism to let users choose freely - this is an addon-on/new feature imo, that shouldnt prevent this bug from getting fixed

**Comment by unkish** [comment]
Safety harbor: expressing personal opinion
<br>

> what does "properly" mean for you here?

A solution that takes into account not only target version but also addresses/minimizes potential breaking change.
I wasn't aware of the motivation behind current solution: _most people don't care about this annotation, they just want something that works without receiving an error on compilation the first time they use jsonschema2pojo_
<br>

If we would assume that most people:

- use JDK 11+ for compiling & running (might not be the case for android devs ?)
- rarely look into all available options when trying out/running `jsonschema2pojo` for the first time, meaning that `targetVersion` stays untouched having default value: `1.6`

With current solution:

1. generating POJOs with JDK 11+ and compiling code for JDK 11+ would succeed (`javax.annotation.processing.Generated` visible to JDK 11+)
2. generating POJOs with JDK 8 and compiling code for JDK's 6-8 would succeed (`javax.annotation.Generated` visible to JDK 8)
3. generating POJOs with JDK 11+ and compiling code for JDK's 6-8 would fail. Workaround: add `jsonschema2pojo-jdk-annotation` or `javax.annotation:javax.annotation-api` dependency

With solution honoring `targetVersion`:

1. would fail as `targetVersion` would be honored (even though it being default one: `1.6`)
2. would succeed
3. would succeed
<br>

It could be argued that changing _default_ value of `targetVersion` from `1.6` to `11` (or higher ?) in addition to honoring target version change would "shift" the failures related to versioning from people who use JDK 11+ and do not target versions 1.8- to people who use JDK 11+ and target versions 1.8- without explicitly overriding `targetVersion`.
In an attempt to lower frictions for new users maybe it would even make sense to revise examples by explicitly adding/mentioning `targetVersion` property.

**Comment by eirnym** [comment]
@unkish this annotation is default in current default settings provided by the library

**Comment by joelittlejohn** [comment]
The hilarious thing is that any tools that check this annotation (e.g. static analysis tools) now have to support all three possible annotations. Some of them are not under particularly active development, so it may take many years before the jakarta one is a viable choice.

**Comment by XN137** [comment]
thanks for bringing up the compatibility concerns... but i think most of them are because targetVersion defaults to 1.6, so my 2 cents:
+1 to determining the default `targetVersion` automatically based on the JVM executing the build

then with my patch, when targeting java 8 one can manually set the property and get the right annotation

**Comment by joelittlejohn** [comment]
Yes, good point @eirnym. We also have the build settings to consider. So the rules for targetVersion would be:

1. If targetVersion is set, use that.
2. If targetVersion is not set, but the build tool has some way of setting target version that we can inspect, use that.
3. If targetVersion is not set, and the build tool has not set a target version either, then use the current JVM version as the target.

Once this is in place, we can choose the annotation based on target version.

**Comment by eirnym** [comment]
@joelittlejohn Good point, I agree with it.

**Comment by XN137** [comment]
https://github.com/joelittlejohn/jsonschema2pojo/blob/8b013e74854fe84f2039ce3d9c9d1874999ccc20/jsonschema2pojo-maven-plugin/src/main/java/org/jsonschema2pojo/maven/Jsonschema2PojoMojo.java#L639-L646

i am no maven expert, but doesnt this mean that 1) and 2) is already respected?

side note:
wondering what happens if you only set the release property (which our build does) as suggested here:
https://maven.apache.org/plugins/maven-compiler-plugin/examples/set-compiler-release.html
will the value be the empty string then? or still `"1.6"` ?

in my PR I could add support for 3) like this:
we change the value of `targetVersion` to `"detect"` then at mojo execution time when it still is `"detect"` we call logic similar to [here](https://github.com/diffplug/spotless/blob/05b87790807c8e8f5f2a0623388492da3bc5d748/lib/src/main/java/com/diffplug/spotless/Jvm.java#L38-L57) (with additional support for 1.6 and 1.7)
and if that also does not give a sane value, we use `"1.6"` as the final fallback.

would this be a good approach?

**Comment by unkish** [comment]
Gradle plugin would have to be adjusted as well.

**Comment by joelittlejohn** [comment]
Yep, I have some changes made yesterday that I will push for review.

> i am no maven expert, but doesnt this mean that 1) and 2) is already respected?

It's a bit more complex, as there are quite a few ways to set this in maven (multiple properties, config on the compiler plug-in).

And I also started looking at gradle, but nothing finished they're yet.

**Comment by Divine1** [comment]
@XN137 


adding below dependency fixed my issue

```
                 <dependency>
			<groupId>jakarta.annotation</groupId>
			<artifactId>jakarta.annotation-api</artifactId>
			<version>2.1.1</version>
		</dependency>
```




this is my pom.xml


```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<parent>
		<groupId>com.sardine.dev</groupId>
		<artifactId>sardine-parent</artifactId>
		<version>0.0.1-SNAPSHOT</version>
	</parent>
	<artifactId>web-model</artifactId>
	<packaging>jar</packaging>

	<properties>
		<spring-boot.repackage.skip>true</spring-boot.repackage.skip>
	</properties>

	<dependencies>

		<!-- https://mvnrepository.com/artifact/org.apache.commons/commons-lang3 -->
		<dependency>
			<groupId>org.apache.commons</groupId>
			<artifactId>commons-lang3</artifactId>
			<version>3.12.0</version>
		</dependency>

		<dependency>
			<groupId>com.fasterxml.jackson.core</groupId>
			<artifactId>jackson-databind</artifactId>
			<version>2.9.7</version>

		</dependency>
		<dependency>
			<groupId>jakarta.annotation</groupId>
			<artifactId>jakarta.annotation-api</artifactId>
			<version>2.1.1</version>
		</dependency>

	</dependencies>

	<build>
		<plugins>
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-compiler-plugin</artifactId>
				<version>3.8.1</version>
				<configuration>
					<source>1.8</source>
					<target>1.8</target>
				</configuration>

			</plugin>
			<plugin>
				<groupId>org.jsonschema2pojo</groupId>
				<artifactId>jsonschema2pojo-maven-plugin</artifactId>
				<version>1.1.3</version>
				<configuration>
					<sourceDirectory>${basedir}/src/main/resources/schema</sourceDirectory>
					<targetPackage>com.sardine.dev.json.model</targetPackage>
					<useCommonsLang3>true</useCommonsLang3>
					<addGeneratedAnnotation>true</addGeneratedAnnotation>
					<additionalProperties>false</additionalProperties>
				</configuration>
				<executions>
					<execution>
						<goals>
							<goal>generate</goal>
						</goals>
					</execution>
				</executions>
			</plugin>
		</plugins>
	</build>
</project>
```

**Comment by XN137** [comment]
@joelittlejohn thanks a lot for addressing this issue so quickly
